Fix mixin initializers

daumayr opened the pull request:
This PR fixes issues with the lookup of mixin initialisers of mixins used in superclasses. 
Previously the following example would not call the initializer of Basic, resulting in getField returning nil. 

```
class Test usingPlatform: platform = Value ()(
  class Basic = (
    | aField ::= false. |
    'basic' println.
    )
  (
    getField = ( ^aField )
  )

  class Super = Object <: Basic (
    'super' println.
  )()

  class Mixin = ()()

  class Sub = Super <: Mixin (
    'Sub' println.
  )(
    public aMethod = (
      getField println.
    )
  )

  public main: args = (
    |s|
    s:: Sub new.
    s aMethod.
    ^ 0
  )
)
```

The mixin initializer were Protected, causing a lookupPrivate call to default to lookupMethod, as the mixin initialisers of Super and Sub have identical names, lookupMethod returns the initializer for Mixin which is found at the beginning of the lookup in the class Sub

The problem was solved by making the mixin initialisers private.

smarr commented on src/som/compiler/MixinBuilder.java:
```diff
@@ -607,8 +611,10 @@ public static SSymbol getInitializerName(final SSymbol selector) {
-    return symbolFor("initializer`" + mixinId + "`" + selector.getString());
+      final int mixinId, final MixinDefinitionId mixinDefinitionId) {
+    return symbolFor(
+        "initializer`" + mixinId + "`" + "_" + mixinDefinitionId + "_"
```
Do we really need both, `mixinId` and `mixinDefinitionId`?

daumayr replied:
mixinId is needed to distinguish multiple mixins used by the same class

smarr replied:
Ah, and please stick with a single style of separators.
For some reason, I decided that the back tick is good, probably, because it can't be part of a normal identifier while `_` can.

smarr commented on src/som/compiler/MixinDefinition.java:
```diff
@@ -431,7 +431,8 @@ private void reportErrorAndExit(final String msgPart1, final String msgPart2) {
   }
 
-  private SInitializer assembleMixinInitializer(final int mixinId) {
+  private SInitializer assembleMixinInitializer(final int mixinId,
```
`mixinId` still needed?

smarr commented on src/som/compiler/MixinDefinition.java:
```diff
@@ -367,7 +367,7 @@ protected boolean hasOnlyImmutableFields(final HashSet<SlotDefinition> instanceS
           }
         }
 
-        SInitializer mixinInit = cdef.assembleMixinInitializer(i);
```
I suspect the for loop above can be changed to be a for each loop over `mixins`, we don't need the `i` anymore,  I think.

smarr reviewed (changes requested):
Great, looks like a clean and simple fix! :)

And might allow for the mentioned simplifications.

Could you also please add tests to:

https://github.com/smarr/SOMns/blob/master/core-lib/TestSuite/MixinTests.ns

Would be good to test a couple of things.
Deeper chains and mixin in the same thing at various levels, and various orders. Ideally tests that fail before applying the fix.

daumayr committed:
Ensure mixin initializers are not overridden by subclasses.

Change visibility of mixin initializers to private. They are only used
by the specific class anyway, and are members of the class. This way, we
avoid them to be overridden by a subclass, which might also have mixins
with the same signature.

daumayr committed:
Added Regression Test

smarr commented:
Reworded the commit message a bit to state the problem.

Thanks for debugging and fixing this!
